src/sp/it/pl/util/async/future/Fut.kt

```diff
@@ -0,0 +1,150 @@
+    fun useBy(executor: Executor = defaultExecutor, block: Consumer<in T>) = then(executor) { block(it); it }
+
+    /**
+     * Returns new future, which sets progress to 0 on fx thread, then executes
+     * this future and then sets progress to 1, again on fx thread.
+     *
+     * Note that when chaining futures, the position within chain decides when
+     * does the progress reach 1. It will not be at the end of the chain, but
+     * at the position of this method in it. The progress is set to 0 always at
+     * the beginning of the computation, i.e. the chain length or position of
+     * this method within it does not have effect.
+     *
+     * To set the progress to 1 at the end of computation, this method must be
+     * the last element of the chain.
+     * To set the progress to 0 somewhere during the computation, a future for
+     * the progress computation must created, this method called on it and
+     * passed as Runnable into another future which executes it as
+     * part of its computation. This will cause only that computation to be bound to
+     * the progress.
+     */
+    fun showProgress(p: Optional<ProgressIndicator>): Fut<T> = p.map { showProgress(it) }.orElse(this)
+
+    /** Invokes [showProgress] using new progress indicator in the currently active window iff a window is active. */
+    fun showProgressOnActiveWindow() = showProgress(APP.windowManager.active.map { it.taskAdd() })
+
+    fun showProgress(p: ProgressIndicator) = Fut<T>(
+            CompletableFuture
+                .runAsync({ p.progress = -1.0 }, FX.kt)
+                .thenComposeAsync { f }
+                .thenApplyAsync({ p.progress = 1.0; it }, FX.kt)
+        )
+
+    /**
+     * @param condition test that if false, progress will not be displayed
+     * @param p function that supplies nonnull progress indicator
+     * @return this (fluent style)
+     */
+    fun showProgress(condition: Boolean, p: () -> ProgressIndicator) = if (condition) showProgress(p()) else this
+
+    fun onOk(executor: Executor = defaultExecutor, block: (T) -> Unit) = onDone(executor) { it.ifOk(block) }
+
+    fun onError(executor: Executor = defaultExecutor, block: (Throwable) -> Unit) = onDone(executor) { it.ifError(block) }
+
+    fun onDone(executor: Executor = defaultExecutor, block: (Try<T,Throwable>) -> Unit): Unit = run {
+        f.handleAsync(
+                { t, e ->  block(if (e==null) Try.ok(t) else Try.error(e)) },
+                executor.kt
+        )
+    }
+
+    fun isDone(): Boolean = f.isDone
+
+    fun getDone(): Try<T,Exception> = try {
+            Try.ok(f.get())
+        } catch (e: InterruptedException) {
+            logger.error(e) { "Asynchronous computation was interrupted" }
+            Try.error(e)
+        } catch (e: ExecutionException) {
+            logger.error(e) { "Asynchronous computation encountered a problem" }
+            Try.error(e)
+        }
+
+    fun getDoneOrNull(): T? = if (f.isDone) {
```

Comment:
```suggestion
    fun getDoneOrNull(): T? = if(f.isDone) getDone().getOr(null) else null
```

Or find another way, but you need to get rid of this code duplication.

Reply:
I do not want to introduce coupling. Duplication isnt is evil only if changing one instance of it, the other should change too. Not the case here.

Out of scope. Larger issue. Added TODO, will do later.